cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java

```diff
@@ -381,121 +276,359 @@ private InetAddress getLocalHost() {
+        LOG.error("{} failed with exception; restarting with back-off", Constants.Service.MASTER_SERVICES, failure);
+        backoffRun();
       }
 
-      // we have to start the dataset service. Because twill services are already running,
-      // it will not be started by the service listener callback below.
-      if (!dsService.isRunning()) {
-        // we need a new dataset service (the old one may have run before and can't be restarted)
-        dsService = baseInjector.getInstance(DatasetService.class); // not a singleton
-        LOG.info("Starting Dataset service");
-        dsService.startAndWait();
+      @Override
+      public void terminated(Service.State from) {
+        if (executor.isShutdown()) {
+          return;
+        }
+        LOG.warn("{} was terminated; restarting with back-off", Constants.Service.MASTER_SERVICES);
+        backoffRun();
       }
 
-    } else {
-      LOG.info("Starting {} application", serviceName);
-      TwillPreparer twillPreparer = getPreparer();
-      twillController = twillPreparer.start();
-
-      twillController.addListener(new ServiceListenerAdapter() {
-
-        @Override
-        public void running() {
-          if (!dsService.isRunning()) {
-            // we need a new dataset service (the old one may have run before and can't be restarted)
-            dsService = baseInjector.getInstance(DatasetService.class); // not a singleton
-            LOG.info("Starting Dataset service");
-            dsService.startAndWait();
-          }
+      private void backoffRun() {
+        if (System.currentTimeMillis() - startTime > SUCCESSFUL_RUN_DURATON_MS) {
+          // Restart immediately
+          executor.execute(new Runnable() {
+            @Override
+            public void run() {
+              monitorTwillApplication(executor, 0, serviceController);
+            }
+          });
+          return;
         }
 
-        @Override
-        public void failed(Service.State from, Throwable failure) {
-          LOG.error("{} failed with exception; restarting with back-off", serviceName, failure);
-          backOffRun();
-        }
+        long nextRunTime = Math.min(500 * (long) Math.pow(2, failures + 1), MAX_BACKOFF_TIME_MS);
+        executor.schedule(new Runnable() {
+          @Override
+          public void run() {
+            monitorTwillApplication(executor, failures + 1, serviceController);
+          }
+        }, nextRunTime, TimeUnit.MILLISECONDS);
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+  }
 
-        @Override
-        public void terminated(Service.State from) {
-          LOG.warn("{} was terminated; restarting with back-off", serviceName);
-          backOffRun();
+  /**
+   * Returns the {@link TwillController} for the current master service or {@code null} if none is running.
+   */
+  @Nullable
+  private TwillController getCurrentTwillController() {
+    int count = 100;
+    long sleepMs = LOOKUP_ATTEMPT_TIMEOUT_MS / count;
```

Comment:
Since this value is actually static final maybe could just move to become static like LOOKUP_ATTEMPT_TIMEOUT_MS ?